File: MMORPG_2/Assets/Script/Data/Proto/CreateProto/RoleInfoRequestProto.cs
//===================================================
//作    者：xxx
//创建时间：2019-05-18 15:13:21
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary> 角色信息(请求) </summary>
public struct RoleInfoRequestProto : IProto
{
    public ushort ProtoCode { get { return 10203; } }

    /// <summary> 角色编号 </summary>
    public int RoleId;

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteInt(RoleId);
            return ms.ToArray();
        }
    }

    public static RoleInfoRequestProto GetProto(byte[] buffer)
    {
        RoleInfoRequestProto proto = new RoleInfoRequestProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.RoleId = ms.ReadInt();
        }
        return proto;
    }
}

File: GameServer/GameServerApp/GameServerApp/Proto/CreateLuaProto/RoleInfoRequestProto.lua
--角色信息(请求)
RoleInfoRequestProto = { ProtoCode = 10203, RoleId = 0 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
RoleInfoRequestProto.__index = RoleInfoRequestProto;

function RoleInfoRequestProto.New()
    local self = { }; --初始化self
    setmetatable(self, RoleInfoRequestProto); --将self的元表设定为Class
    return self;
end


--发送协议
function RoleInfoRequestProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.RoleId);

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function RoleInfoRequestProto.GetProto(buffer)

    local proto = RoleInfoRequestProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.RoleId = ms:ReadInt();

    ms:Dispose();
    return proto;
end